superset/db_engine_specs.py

```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+        df = pandas.DataFrame()
+        df = pandas.concat(chunk for chunk in chunks)
+        return df
+
+    @staticmethod
+    def df_to_db(df, name, con, schema, if_exists, index,
+                 index_label, chunksize):
+        engine = create_engine(con, echo=False)
+        # Use Pandas to parse dataframe to database
+        df.to_sql(name=name, con=engine, schema=schema, if_exists=if_exists,
+                  index=index, index_label=index_label, chunksize=chunksize)
+
+        database = (
+            db.session
+                .query(type(query)) #Database
+                .filter_by(sqlalchemy_uri=con)
+                .first()
+        )
+        table.database_id = database.id
+        table.user_id = g.user.id
+        table.database = database
+        table.schema = form.schema.data
+        db.session.add(table)
+        db.session.commit()
+
+        #post_add
+        table.fetch_metadata()
+        security.merge_perm(sm, 'datasource_access', table.get_perm())
+        if table.schema:
+            security.merge_perm(sm, 'schema_access', table.schema_perm)
+
+
+    @staticmethod
+    def upload_csv(form, table):
+        def allowed_file(filename):
+            # Only allow specific file extensions as specified in the config
+            extension = os.path.splitext(filename)[1]
+            return extension and extension[1:] in config['ALLOWED_EXTENSIONS']
+
+        filename = secure_filename(form.csv_file.data.filename)
+        if not allowed_file(filename):
+            flash("Invalid file trype selected.", 'alert')
+            return False
+        df = BaseEngineSpec.csv_to_df(names=form.names.data,
+                            filepath_or_buffer=filename,
+                            sep=form.sep.data,
+                            header=form.header.data,
+                            index_col=form.index_col.data,
+                            squeeze=form.squeeze.data,
+                            prefix=form.prefix.data,
+                            mangle_dupe_cols=form.mangle_dupe_cols.data,
+                            skipinitialspace=form.skipinitialspace.data,
+                            skiprows=form.skiprows.data,
+                            nrows=form.nrows.data,
+                            skip_blank_lines=form.skip_blank_lines.data,
+                            parse_dates=form.parse_dates.data,
+                            infer_datetime_format=form.infer_datetime_format.data,
+                            dayfirst=form.dayfirst.data,
+                            thousands=form.thousands.data,
+                            decimal=form.decimal.data,
+                            quotechar=form.quotechar.data,
+                            escapechar=form.escapechar.data,
+                            comment=form.comment.data,
+                            error_bad_lines=form.error_bad_lines.data,
+                            chunksize=10000)
+
+        BaseEngineSpec.df_to_db(df=df,
```

Comment:
NIT: break after `(` is your linter not triggering on this?

Reply:
Fixed